src/messages/helpers/receive.rs

```diff
@@ -0,0 +1,188 @@
+    Error,
+    Jwe,
+    Jws,
+    Message,
+    MessageType,
+    Recipient,
+    Signature,
+};
+
+/// Helper type to check if received message is plain, signed or encrypted
+#[derive(Serialize, Deserialize, Debug)]
+struct UnknownReceivedMessage<'a> {
+    #[serde(borrow)]
+    pub signature: Option<&'a RawValue>,
+
+    #[serde(borrow)]
+    pub signatures: Option<&'a RawValue>,
+
+    #[serde(borrow)]
+    pub iv: Option<&'a RawValue>,
+}
+
+/// Tries to parse message and checks for well known fields to derive message type.
+pub(crate) fn get_message_type(message: &str) -> Result<MessageType, Error> {
+    // try to skip parsing by using known fields from jwe/jws
+    let to_check: UnknownReceivedMessage = serde_json::from_str(message)?;
+    if to_check.iv.is_some() {
+        return Ok(MessageType::DidCommJwe);
+    }
+    if to_check.signatures.is_some() || to_check.signature.is_some() {
+        return Ok(MessageType::DidCommJws);
+    }
+    let message: Message = serde_json::from_str(message)?;
+    Ok(message.jwm_header.typ)
+}
+
+/// Receive a serialized message. This function handles receival of [`crate::Jwe`] envelopes.
+///
+/// # Arguments
+///
+/// * `incoming` - incoming message
+///
+/// * `encryption_recipient_private_key` - private key of recipient of a message, required
+///
+/// * `encryption_sender_public_key` - public key of message sender, can be omitted if public key
+///                                    should be automatically resolved (requires `resolve` feature)
+pub(crate) fn receive_jwe(
+    incoming: &str,
+    encryption_recipient_private_key: &[u8],
+    encryption_sender_public_key: Option<&[u8]>,
+) -> Result<String, Error> {
+    let jwe: Jwe = serde_json::from_str(incoming)?;
+    let alg = &jwe
+        .get_alg()
+        .ok_or_else(|| Error::Generic("missing algorithm in JWE header(s)".to_string()))?;
+
+    // get public key from input or from senders DID document
+    let sender_public_key = match encryption_sender_public_key {
+        Some(value) => value.to_vec(),
+        None => {
+            #[cfg(feature = "resolve")]
+            {
+                let skid = &jwe
+                    .get_skid()
+                    .ok_or_else(|| Error::Generic("skid missing".to_string()))?;
+                let document = ddoresolver_rs::resolve_any(skid).ok_or(Error::DidResolveFailed)?;
+                document
+                    .find_public_key_for_curve("X25519")
+                    .ok_or(Error::BadDid)?
+            }
+            #[cfg(not(feature = "resolve"))]
+            {
+                return Err(Error::DidResolveFailed);
+            }
+        }
+    };
+
+    let shared = StaticSecret::from(array_ref!(encryption_recipient_private_key, 0, 32).to_owned())
+        .diffie_hellman(&PublicKey::from(
+            array_ref!(sender_public_key, 0, 32).to_owned(),
+        ));
+    let a: CryptoAlgorithm = alg.try_into()?;
+    let m: Message;
```

Comment:
I would suggest to move `a` and `m` a bit down before the `if let Some(recipients)` block (and give them some better names) but keep the definition on the same scope.

Also addressing the next comment, removing the definition and serializing the values in side the `if` would look more or less like this:

```rs
        Ok(serde_json::to_string(&Message::decrypt(
            incoming.as_bytes(),
            a.decrypter(),
            &key,
        )?)?)
    } else {
        Ok(serde_json::to_string(&Message::decrypt(
            incoming.as_bytes(),
            a.decrypter(),
            shared.as_bytes(),
        )?)?)
    }
```

instead of the current

```rs
        m = Message::decrypt(incoming.as_bytes(), a.decrypter(), &key)?;
    } else {
        m = Message::decrypt(incoming.as_bytes(), a.decrypter(), shared.as_bytes())?;
    }

    Ok(serde_json::to_string(&m)?)
```

In my opinion the current version has less returns in some if/else blocks and avoids duplicated code for the serialization.

Buf if you are in favor of removing the `m` I would also comply. ^^

Reply:
Let's leave it the way it's now, then:)